### PR TITLE
Add success and failure for Plugin metrics

### PR DESCRIPTION
**Pull request by anandswaminathan**
In place of https://github.com/lyft/flyteplugins/pull/93

### CODE REVIEW

File: pkg/controller/nodes/task/handler.go [hunk 1]
```diff
@@ -189,6 +195,12 @@ func (t *Handler) Setup(ctx context.Context, sCtx handler.SetupContext) error {
 		for _, tt := range p.RegisteredTaskTypes {
 			logger.Infof(ctx, "Plugin [%s] registered for TaskType [%s]", p.ID, tt)
 			t.plugins[tt] = cp
+			t.taskMetricsMap[tt] = &taskMetrics{
```
Comment (EngHabu) [inline pkg/controller/nodes/task/handler.go:198]: Can you add the plugin type as well in the map key and the metric name too? We are designing how to allow multiple plugins for the same task type and choose the right one at runtime as we speak...

File: pkg/controller/nodes/task/handler.go [hunk 2]
```diff
@@ -189,6 +195,12 @@ func (t *Handler) Setup(ctx context.Context, sCtx handler.SetupContext) error {
...
+				taskSucceeded: labeled.NewCounter(tt+"_success",
```
Comment (EngHabu) [inline pkg/controller/nodes/task/handler.go:199]: I don't think we currently have restrictions on how to name task types... but I know prometheus does have restrictions on the character set... you might want to sanitize them before using them? I could be wrong

Comment (anandswaminathan) [reply pkg/controller/nodes/task/handler.go]: done

File: pkg/controller/nodes/task/handler.go (hunk 1 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/task/handler.go]: I am unable to understand plugin type. Have not addressed.

File: pkg/controller/nodes/task/handler.go [hunk 3]
```diff
@@ -189,6 +198,18 @@ func (t *Handler) Setup(ctx context.Context, sCtx handler.SetupContext) error {
...
 			logger.Infof(ctx, "Plugin [%s] registered for TaskType [%s]", p.ID, tt)
 			t.plugins[tt] = cp
+			metricName, err := utils.GetSanitizedPrometheusKey(tt)
+			if err != nil {
+				return err
+			}
+			if _, ok := t.taskMetricsMap[metricName]; !ok {
+				t.taskMetricsMap[metricName] = &taskMetrics{
+					taskSucceeded: labeled.NewCounter(metricName+"_success",
+						"Task finished successfully", t.pluginScope, labeled.EmitUnlabeledMetric),
```
Comment (kumare3) [inline pkg/controller/nodes/task/handler.go:208]: Can we add the tasktype in description?

File: pkg/controller/nodes/task/handler.go [hunk 4]
```diff
@@ -298,6 +319,15 @@ func (t Handler) invokePlugin(ctx context.Context, p pluginCore.Plugin, tCtx *ta
 		}
 	}
 
+	if !pluginTrns.IsPreviouslyObserved() {
```
Comment (kumare3) [inline pkg/controller/nodes/task/handler.go:322]: should we also make this `t.metrics.pluginPanics.Inc(ctx)` a labeled metric?

File: pkg/utils/helpers.go [hunk 5]
```diff
@@ -10,3 +12,12 @@ func CopyMap(o map[string]string) (r map[string]string) {
 	}
 	return
 }
+
+func GetSanitizedPrometheusKey(key string) (string, error) {
+	// Make a Regex to say we only want letters and numbers
```
Comment (kumare3) [inline pkg/utils/helpers.go:17]: letters, numbers, `_` and `:`

**Comment by kumare3** [comment]
Couple comments, but this is great. We should get a common way of generating metrics for all of them

File: pkg/controller/nodes/task/handler.go (hunk 3 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/task/handler.go:208]: done

File: pkg/controller/nodes/task/handler.go (hunk 4 above)
Comment (anandswaminathan) [reply pkg/controller/nodes/task/handler.go:322]: It is but not tagged at task type level.

File: pkg/controller/nodes/task/handler_test.go [hunk 6]
```diff
@@ -806,8 +815,17 @@ func Test_task_Handle_Catalog(t *testing.T) {
...
 			assert.NoError(t, err)
+			fakePlugin := fakeplugins.NewPhaseBasedPlugin()
 			tk.plugins = map[pluginCore.TaskType]pluginCore.Plugin{
-				"test": fakeplugins.NewPhaseBasedPlugin(),
+				"test": fakePlugin,
+			}
+			tk.taskMetricsMap = map[pluginCore.TaskType]*taskMetrics{
+				"test_" + fakePlugin.GetID(): &taskMetrics{
+					taskSucceeded: labeled.NewCounter("success",
+						"Task finished successfully", promutils.NewTestScope(), labeled.EmitUnlabeledMetric),
```
Comment (kumare3) [inline pkg/controller/nodes/task/handler_test.go:825]: Hey I do not see my comment. Can we reword the description to have the tasktype information.

**Comment by anandswaminathan** [comment]
This is test file.

File: pkg/controller/nodes/task/handler.go [hunk 7]
```diff
@@ -59,6 +66,9 @@ type pluginRequestedTransition struct {
 	pluginStateVersion uint32
 }
 
+func getPluginMetricKey(taskType string, id string) string {
```
Comment (EngHabu) [inline pkg/controller/nodes/task/handler.go:69]:
```suggestion
func getPluginMetricKey(pluginID, taskType string) string {
```

File: pkg/controller/nodes/task/handler.go [hunk 8]
```diff
@@ -59,6 +66,9 @@ type pluginRequestedTransition struct {
...
+	return taskType + "_" + id
```
Comment (EngHabu) [inline pkg/controller/nodes/task/handler.go:70]:
```suggestion
	return taskType + "_" + pluginID
```

File: pkg/controller/nodes/task/handler.go [hunk 9]
```diff
@@ -229,6 +241,22 @@ func validateTransition(transition pluginCore.Transition) error {
 	return nil
 }
 
+func (t Handler) fetchPluginTaskMetrics(taskType, pluginID string) (*taskMetrics, error) {
+	metricNameKey, err := utils.GetSanitizedPrometheusKey(getPluginMetricKey(taskType, pluginID))
```
Comment (EngHabu) [inline pkg/controller/nodes/task/handler.go:245]:
```suggestion
	metricNameKey, err := utils.GetSanitizedPrometheusKey(getPluginMetricKey(pluginID, taskType))
```

File: pkg/controller/nodes/task/handler.go [hunk 10]
```diff
@@ -598,6 +640,7 @@ func New(ctx context.Context, kubeClient executors.Client, client catalog.Client
 	return &Handler{
 		pluginRegistry: pluginMachinery.PluginRegistry(),
 		plugins:        make(map[pluginCore.TaskType]pluginCore.Plugin),
+		taskMetricsMap: make(map[pluginCore.TaskType]*taskMetrics),
```
Comment (EngHabu) [inline pkg/controller/nodes/task/handler.go:643]:
```suggestion
		taskMetricsMap: make(map[string]*taskMetrics),
```

or a new "MetricKey" type alias?

File: pkg/controller/nodes/task/handler.go [hunk 11]
```diff
@@ -133,6 +143,7 @@ type Handler struct {
 	catalog         catalog.Client
 	asyncCatalog    catalog.AsyncClient
 	plugins         map[pluginCore.TaskType]pluginCore.Plugin
+	taskMetricsMap  map[pluginCore.TaskType]*taskMetrics
```
Comment (EngHabu) [inline pkg/controller/nodes/task/handler.go:146]:
```suggestion
	taskMetricsMap  map[string]*taskMetrics
```

or a new `MetricKey` type alias?

**Pull request merged**